moshi/src/main/java/com/squareup/moshi/JsonValueReader.kt

```diff
@@ -0,0 +1,391 @@
+    } else if (stackSize > 0) {
+      // We're skipping a value.
+      remove()
+    } else {
+      throw JsonDataException("Expected a value but was ${peek()} at path $path")
+    }
+  }
+
+  override fun nextSource(): BufferedSource {
+    val value = readJsonValue()
+    val result = Buffer()
+    JsonWriter.of(result).use { jsonWriter -> jsonWriter.jsonValue(value) }
+    return result
+  }
+
+  override fun peekJson(): JsonReader = JsonValueReader(this)
+
+  override fun promoteNameToValue() {
+    if (hasNext()) {
+      val name = nextName()
+      push(name)
+    }
+  }
+
+  override fun close() {
+    stack.fill(null, 0, stackSize)
+    stack[0] = JSON_READER_CLOSED
+    scopes[0] = JsonScope.CLOSED
+    stackSize = 1
+  }
+
+  private fun push(newTop: Any?) {
+    if (stackSize == stack.size) {
+      if (stackSize == 256) {
+        throw JsonDataException("Nesting too deep at $path")
+      }
+      scopes = scopes.copyOf(scopes.size * 2)
+      pathNames = pathNames.copyOf(pathNames.size * 2)
+      pathIndices = pathIndices.copyOf(pathIndices.size * 2)
+      stack = stack.copyOf(stack.size * 2)
+    }
+    stack[stackSize++] = newTop
+  }
+
+  private inline fun <reified T> require(expected: Token): T = knownNotNull(require(T::class.java, expected))
+
+  private fun requireNull() = require(Void::class.java, Token.NULL)
+
+  /**
+   * Returns the top of the stack which is required to be a `type`. Throws if this reader is
+   * closed, or if the type isn't what was expected.
+   */
+  private fun <T> require(type: Class<T>, expected: Token): T? {
+    val peeked = if (stackSize != 0) stack[stackSize - 1] else null
+    if (type.isInstance(peeked)) {
+      return type.cast(peeked)
+    }
+    if (peeked == null && expected == Token.NULL) {
+      return null
+    }
+    ifNotClosed(peeked) {
+      throw typeMismatch(peeked, expected)
+    }
+  }
+
+  private fun stringKey(entry: Map.Entry<*, *>): String {
+    val name = entry.key
+    if (name is String) return name
+    throw typeMismatch(name, Token.NAME)
+  }
+
+  private inline fun <T> ifNotClosed(peeked: Any?, body: () -> T): T {
+    check(peeked !== JSON_READER_CLOSED) { "JsonReader is closed" }
+    return body()
+  }
+
+  /**
+   * Removes a value and prepares for the next. If we're iterating a map or list this advances the
+   * iterator.
+   */
+  private fun remove() {
+    stackSize--
+    stack[stackSize] = null
+    scopes[stackSize] = 0
+
+    // If we're iterating an array or an object push its next element on to the stack.
+    if (stackSize > 0) {
+      pathIndices[stackSize - 1]++
+      val parent = stack[stackSize - 1]
+      if (parent is Iterator<*> && parent.hasNext()) {
+        push(parent.next())
+      }
+    }
+  }
+
+  internal class JsonIterator(
+    val endToken: Token,
+    val array: Array<Any?>,
+    var next: Int
+  ) : Iterator<Any?>, Cloneable {
+    override fun hasNext() = next < array.size
+
+    override fun next() = array[next++]
+
```

Comment:
does it matter that `remove` isn't overridden with an `UnsupportedOperationException`?

Reply:
In this case, no since we're using an immutable interface